src/dos/program_more_output.cpp

```diff
@@ -281,33 +396,170 @@ UserDecision MoreOutputBase::PromptUser()
+	}
+
+	return decision;
+}
+
+MoreOutputBase::UserDecision MoreOutputBase::PromptUserIfNeeded()
+{
+	if (shutdown_requested) {
+		return UserDecision::Cancel;
+	}
+	if (screen_line_counter >= lines_to_display) {
+		return PromptUser();
+	}
+	return UserDecision::More;
+}
+
+uint32_t MoreOutputBase::GetNumLinesFromUser(UserDecision& decision)
+{
+	constexpr size_t max_digits = 5;
+
+	WriteOut(MSG_Get("PROGRAM_MORE_HOW_MANY_LINES"));
+	WriteOut(" ");
+
+	std::string number_str = {};
+	while (!shutdown_requested) {
+		CALLBACK_Idle();
+
+		// Try to read the key
+		uint16_t count = 1;
+		uint8_t code   = 0;
+		DOS_ReadFile(STDIN, &code, &count);
+
+		if (count == 0 || code == code_ctrl_c) {
+			// Terminate the whole displaying
+			decision = UserDecision::Cancel;
+			break;
+		} else if (code == code_esc) {
+			// User has resigned, no number of lines
+			break;
+		} else if (code == code_cr && !number_str.empty()) {
+			// ENTER pressed, we have a valid number
+			return static_cast<uint32_t>(std::stoi(number_str));
+		} else if (code == code_bs && !number_str.empty()) {
+			// BACKSPACE pressed
+			WriteOut("%c %c", code, code);
+			number_str.pop_back();
+		} else if (std::isdigit(code) && number_str.length() < max_digits &&
+		           !(code == '0' && number_str.empty())) {
+			// Add a new digit to the number
+			WriteOut("%c", code);
+			number_str.push_back(static_cast<char>(code));
+		}
+	}
+
+	return 0;
+}
+
+MoreOutputBase::UserDecision MoreOutputBase::WaitForCancelContinue()
+{
+	auto decision = UserDecision::NextFile;
+	while (decision == UserDecision::NextFile) {
+		decision = WaitForCancelContinueNext();
+	}
+
+	return decision;
+}
+
+MoreOutputBase::UserDecision MoreOutputBase::WaitForCancelContinueNext()
+{
+	auto decision = UserDecision::Cancel;
+	while (!shutdown_requested) {
+		CALLBACK_Idle();
+
+		// Try to read the key
+		uint16_t count = 1;
+		uint8_t code   = 0;
+		DOS_ReadFile(STDIN, &code, &count);
+
+		if (shutdown_requested || count == 0 ||
+		    code == 'q' || code == 'Q' ||
+		    code == code_ctrl_c || code == code_esc) {
+			decision = UserDecision::Cancel;
+			break;
+		} else if (code == ' ') {
+			decision = UserDecision::More;
+			break;
+		} else if (code == code_cr) {
+			if (has_option_expand_form_feed) {
+				decision = UserDecision::More;
+			} else {
+				decision = UserDecision::MoreOneLine;
+			}
+			break;
+		} else if (code == 'n' || code == 'N' || // FreeDOS hotkey
+		           code == 'f' || code == 'F') { // Windows hotkey
+			decision = UserDecision::NextFile;
+			break;
+		} else if (has_option_extended_mode) {
+			if (code == '=') {
+				decision = UserDecision::SwitchPrompt;
+				break;
+			} else if (!has_option_expand_form_feed &&
+			           (code == 'p' || code == 'P')) {
+				decision = UserDecision::MoreNumLines;
+				break;
+			} else if (!has_option_expand_form_feed &&
+			           (code == 's' || code == 'S')) {
+				decision = UserDecision::SkipNumLines;
```

Comment:
We've got a char case-insensitive equals operation too (also in `string_utils`:

```c++
string_utils.h:bool ciequals(const char a, const char b);
```

Reply:
Ok, changed - unfortunately, ad the cost of one `static_cast` from `uint8_t` to `char`.